NuRadioReco/modules/channelGalacticNoiseAdder.py

```diff
@@ -160,6 +167,12 @@ def begin(
             self.__radio_sky = healpy.pixelfunc.ud_grade(self.__radio_sky, self.__n_side)
             self.__noise_temperatures[i_freq] = self.__radio_sky
 
+
+    @functools.lru_cache(maxsize=1024 * 32)
```

Comment:
I guess for the galacticnoiseadder you need a large cache... so what is this, 32000 * len(frequencies) * 2 (polarizations) 128-bit complex numbers? So you would typically increase the memory usage by O(1--2 GB), is that right?

Reply:
> Also, I'm pretty sure self is not hashed, so you may want to clear the cache on class initialization as well.

I can not follow.

Reply:
`maxsize` refers to the number of entries in the internal dict which is used for caching not the actual size of the cache. I still read somewhere that it should be a power of two... Right now we allow for ~32k entries which should be by far sufficient.

Reply:
My bad for making two points in a single thread. 32000 = 32 * 1024 (close enough). Each entry contains len(frequencies) * len(polarizations) complex numbers. My rough estimate was this is an additional 1-2 GB of memory, assuming 1000-2000 frequency bins, and this could of course get worse for longer traces or large (upsampled) sampling rates. I wondered if you had checked this.

Reply:
Based on how many entries is that?

Reply:
> > Also, I'm pretty sure self is not hashed, so you may want to clear the cache on class initialization as well.
> 
> I can not follow.

The cache does not reset if you initialize a second instance of the channelGalacticNoiseAdder. It probably should? Although it seems you are also checking if the antenna_pattern has changed... so I guess in that case it might not matter. (As an aside, I didn't know you could hash custom classes inside lru_cache, but I guess you've tested this and it does not complain).

Reply:
Yes I did. And `self.__hash__()` does work so self is hashed. A new instance gets a new hash. Of course the hash does not change when you modify the member variables. But this we keep track of our selfs.

Reply:
So a vector with 2000 entries and dtype "complex128` gives me a size of 32kb. Taking into account 2 polarisations and 80 different directions from the sky give us 5 MB. That should be fine right? It is hard to see for me that this would increase by a factor of ~100

Reply:
If you want 80 directions then why do you make the cache size 32000 instead of 80? I really don't understand where our communication is breaking down at the moment...

Reply:
I do not think this cache size is allocated when not used. It is only the limit of what it is allowed to allocate. A maximum of 2 GB (which is probably never reached) seems absolutely acceptable for me. I set this number while not over thinking it. Feel free to change it if that is what bothers you.

Reply:
I would change it to e.g. 1024. If the intention is for it to use 2 GB then that's a feature that we can discuss. If that's not the intention I'd rather not accidentally use 2 GB of memory (e.g. in a 'realistic' simulation of a detector at different times with different antenna orientations, I see no reason why this wouldn't fill the cache).

Reply:
Well if you save it is realistic to fill the cache up to 2GB I would leave it as it. I think it is more problematic that people do not realise that they fill up the cache and thus suffer from reduced performance as that 2 GB are becoming a problem. (Changing the station time should hardly effect the cache).

Reply:
Changing station time affects the cache - you will ask for different local zeniths and azimuths because the local coordinate system changes as a function of time. In this case you will be asking for new vector effective lengths constantly while storing (unused) vector effective lengths in the cache. This is just a waste of memory for - as far as I can tell - no reason.

You could say "well, just don't use the caching in that case". Or you could just reduce the cache size so there's no downside to using it. I would advocate the latter.

Reply:
No, I would say if this is actually the case we should change what and keep the zenith and azimuth angle in the local  coordinate system fix and only change the sky temperature for the new position in the sky. This will be much more efficient.

Reply:
Or would that just move the problem back to the interpolation of the sky map?

Reply:
Okay sjoerd. YOU WIN.